contracts/EthProxy.sol

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.6.7;
+
+import "@openzeppelin/contracts/utils/Address.sol";
+import "./IUSM.sol";
+import "./external/IWETH9.sol";
+
+
+contract EthProxy {
+    using Address for address payable;
+    IUSM public usm;
+    IWETH9 public weth;
+
+    constructor(address usm_, address weth_) public {
+        usm = IUSM(usm_); 
+        weth = IWETH9(weth_);
+
+        weth.approve(address(usm), uint(-1));
+    }
+
+    /// @dev The WETH9 contract will send ether to EthProxy on `weth.withdraw` using this function.
+    receive() external payable { }
+
+    /// @dev Users use `mint` in EthProxy to post ETH to USM (amount = msg.value), which will be converted to Weth here.
+    /// @param to Receiver of the minted USM
+    function mint(address to)
+        external payable {
+        weth.deposit{ value: msg.value }();
+        usm.mint(address(this), to, msg.value);
```

Comment:
Oy vey, the more closely I read this class the more confused I get!  😂  Probably just because this is all so new to me.  But: above we call `weth.approve(address(usm))`, and here we call `usm.mint(address(this), ...)`...  I'm trying to understand why that doesn't read `usm.mint(address(usm), ...)`.  Does `usm.mint()` end up transferring the WETH from `this`, or from `usm`?  Or does some magic make them the same.

Reply:
`EthProxy` acts as a buffer here.

The user sends ether to `EthProxy` using `proxy.mint(receiver, { value: amount })`.
`EthProxy` converts the ether to weth, in his own account, using `weth.deposit({ value: amount })`
Now that `EthProxy` has weth, it uses it to mint USM in the chosen USM account: `usm.mint(from, to, amount)` -> `usm.mint(address(this), to, amount)` -> `usm.mint(proxy.address, to, amount)`

You might be confused by the existence of `from` and `to` parameters in all functions, that allows to use any function as a means of payment.

The most common use case would be `proxy.mint(user1, { from: user1, value: amount})`

Reply:
Yeah I understand _most_ of that.  My specific confusion is, does the `weth.deposit()` call result in the WETH being added to `address(weth)`, or to `address(this)`?  (Since the `usm.mint(address(this), ...)` call is spending it from `this`.)  Or does some magic mean adding WETH to `weth` is identical to adding it to `this`.

Reply:
Maybe my confusion is that `weth.deposit()` doesn't mean "Add WETH to the `weth` address", it means "Use `weth` to add WETH to this contract's address"?

Reply:
`weth.deposit()` means "Hey, weth contract, take my ETH and give me Weth". `Proxy` takes the ETH from the user, calls `weth.deposit` which credits Weth to the `Proxy` address, and then `Proxy` calls `usm.mint(address(this), to, amount)` to mint USM with its Weth, into the `to` account that the user specified.

So, for an instant, it's `Proxy` that has Weth, which then uses immediately to mint USM in favour of `to`.

Reply:
Yup, OK, that's what my last comment referred to.  I had read `weth.deposit()` as "Deposit into `weth`", but in fact it's "Deposit into me."  So it all makes sense to me now, thanks.